src/controller/CHIPDeviceController.cpp

```diff
@@ -995,5 +1008,95 @@ void DeviceCommissioner::OnSessionEstablishmentTimeoutCallback(System::Layer * a
+    NodeId deviceId                           = reinterpret_cast<chip::NodeId>(static_cast<const app::Command *>(apCommandSender));
+    app::CHIPDeviceCallbacksMgr * callbackMgr = &app::CHIPDeviceCallbacksMgr::GetInstance();
+
+    Callback::Cancelable * onSuccessCallback = nullptr;
+    Callback::Cancelable * onFailureCallback = nullptr;
+
+    // The command response in IM is identified by the command sender pointer -- the response of the command will return to the same
+    // object used for send the command.
+    callbackMgr->GetResponseCallback(deviceId, 0, &onSuccessCallback, &onFailureCallback);
+
+    typedef void (*DefaultFailureCallback)(void * context, uint8_t status);
+
+    // We cannot found the response of the command we just send.
+    VerifyOrReturnError(onSuccessCallback != nullptr && onFailureCallback != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    Callback::Callback<DefaultFailureCallback> * cb = Callback::Callback<DefaultFailureCallback>::FromCancelable(onFailureCallback);
+    // Generally IM has more detailed errors than ember library, here we always use 1 temporary, the actual handling of the commands
+    // should implement full IMDelegate.
+    cb->mCall(cb->mContext, static_cast<uint8_t>(1));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeviceControllerInteractionModelDelegate::CommandResponseProcessed(const app::CommandSender * apCommandSender)
+{
+    // This is a bit tricky, we try to assume that chip::NodeId is uint64_t so the pointer can be used as a NodeId for CallbackMgr.
+    static_assert(std::is_same<chip::NodeId, uint64_t>::value, "chip::NodeId is not uint64_t");
+    NodeId deviceId                           = reinterpret_cast<chip::NodeId>(static_cast<const app::Command *>(apCommandSender));
+    app::CHIPDeviceCallbacksMgr * callbackMgr = &app::CHIPDeviceCallbacksMgr::GetInstance();
+
+    Callback::Cancelable * onSuccessCallback = nullptr;
+    Callback::Cancelable * onFailureCallback = nullptr;
+
+    // The command response in IM is identified by the command sender pointer -- the response of the command will return to the same
+    // object used for send the command.
+    callbackMgr->GetResponseCallback(deviceId, 0, &onSuccessCallback, &onFailureCallback);
+
+    typedef void (*DefaultSuccessCallback)(void * context);
+
+    // We cannot found the response of the command we just send. We prpbably just processed the response of the response. So we
+    // don't consider it as an error.
+    VerifyOrReturnError(onSuccessCallback != nullptr && onFailureCallback != nullptr, CHIP_NO_ERROR);
+
+    // When we can get the callback, that means we did not process this command before, this should be caused by the response of the
+    // command is another command.
+    Callback::Callback<DefaultSuccessCallback> * cb = Callback::Callback<DefaultSuccessCallback>::FromCancelable(onSuccessCallback);
+    cb->mCall(cb->mContext);
+
+    return CHIP_NO_ERROR;
+}
+
```

Comment:
Why do we need this code ? Can't we just used `CHIPClientCallbacks` at that point ?

Reply:
The status code does not goes into the CHIPClientCallbacks.

Reply:
Why is it not going into `CHIPClientCallbacks` ? Can we get all the handlers there ? I don't mind at all if this code is rewritten but it may be easier to follow if there is only a single call site that handles responses.

Reply:
The main reason is that this is not ember library.

Reply:
Sorry I still do not understand.

Reply:
The `IMClusterCommandHandler.cpp` just dispatches the commands, like what we do in `call-command-handler.cpp`

Just added `IMDefaultResponseCallback` to `CHIPClientCallbacks`, now all callback are called in `CHIPClientCallbacks`.